src/sentry/grouping/enhancer/__init__.py

```diff
@@ -138,19 +147,28 @@ def apply_modifications_to_frame(
         """
         in_memory_cache: dict[str, str] = {}
 
-        match_frames = [create_match_frame(frame, platform) for frame in frames]
-
-        with sentry_sdk.start_span(
-            op="stacktrace_processing",
-            description="apply_rules_to_frames",
-        ):
+        match_frames, stacktrace_fingerprint = _matching_frames_and_fingerprint(frames, platform)
+        rules_fingerprint = _generate_rules_fingerprint(self._modifier_rules, platform)
+        # The most expensive part of creating groups is applying the rules to frames (next code block)
+        # We include the rules fingerprint to make sure that the set of rules are still the same
+        cache_key = f"stacktrace_rules_fingerprint.{rules_fingerprint}.{stacktrace_fingerprint}"
+        if stacktrace_fingerprint and rules_fingerprint:
+            merged, merged_frames = _merge_cached_values(frames, cache_key, platform)
+            if merged:
+                frames = merged_frames
+                return
```

Comment:
In order to test out this PR I've added the four lines above. We could decide to merge this PR with the following change just to see the code being exercised but not using the results. That way we can see any hashing issues show up on Sentry and address them before going fully live.
```diff
-            merged, merged_frames = _merge_cached_values(frames, cache_key, platform)
-            if merged:
-                frames = merged_frames
-                return
+            _merge_cached_values(frames, cache_key, platform)
```